File: rom_app/restaurant_ops_mgmt/report/chef_production_register/chef_production_register.js
frappe.query_reports["Chef Production Register"] = {
"filters": [
		{
			"fieldname": "from_date_filter",
			"label": "From Date *",
			"fieldtype": "Date",
			"default": frappe.datetime.add_days(frappe.datetime.get_today(), -1),
			"mandatory": 1,
		},
		{
			"fieldname": "to_date_filter",
			"label": "To Date *",
			"fieldtype": "Date",
			"default": frappe.datetime.now_date(),
			"mandatory": 1,
		},
		{
			"fieldname": "branch_filter",
			"label": "Branch",
			"fieldtype": "Link",
			"options": "Branch",
		},
		{
			"fieldname": "category_filter",
			"label": "Category",
			"fieldtype": "Select",
			"options": "\nBriyani\nChicken"
		},
		{
			"fieldname": "item_filter",
			"label": "Item",
			"fieldtype": "Select",
		},
	],
	onload:function(){
		console.log('onload');
		var select_item  = $("select[data-fieldname='item_filter']");

		let urlParams = new URLSearchParams(window.location.search);
		let item_filter_value = urlParams.get('item_filter');

		if(item_filter_value){
			console.log('item_filter_value', item_filter_value);
			select_item.append('<option value="" ></option>');
		}
		else
		{
			select_item.append('<option value="" selected="selected"></option>');
		}
		PopulateDropDownList();
	},

};

function PopulateDropDownList() {
	var ddlItems  = $("select[data-fieldname='item_filter']");
	let api_url = "rom_app.restaurant_ops_mgmt.api.get_production_items"

	frappe.call({
		method: api_url,
		callback: function(items) {

			console.log('get_production_items');
			console.log(items);
			$.each(items.message, function(_i, e){
				console.log('_i',_i);
				console.log('e',e);
				let each_item = e[0];
				console.log('=======> ',each_item);

				var option;

				let urlParams = new URLSearchParams(window.location.search);
				let item_filter_value = urlParams.get('item_filter');

				if(item_filter_value == each_item){
					console.log('item_filter_value', item_filter_value);
					option = $('<option selected="selected" />');
				}
				else
				{
					console.log('else item_filter_value');
					option = $('<option />');
				}

				option.html(each_item);
				option.val(each_item);
				ddlItems.append(option);
			});
		}
	});
};


/*
function PopulateDropDownList() {
	console.log("PopulateDropDownList");
	var ddlItems  = $("select[data-fieldname='item_filter']");
	let api_url = "rom_app.restaurant_ops_mgmt.api.get_production_items"

	frappe.call({
		method: api_url,
		callback: function(items) {
			if(items){



				console.log('get_production_items');
				console.log(items);
				$.each(items.message, function(_i, e){
					console.log('_i',_i);
					console.log('e',e);
					let each_item = e[0];
					console.log('=======> ',each_item);


					var option = $("<option />");
					option.html(each_item);
					option.val(each_item);

					ddlItems.append(option);
				});

				//check if the url has any paramter and then set it as default
				// let urlParams = new URLSearchParams(window.location.search);
				// let item_filter_value = urlParams.get('item_filter');
    //
				// if(item_filter_value){
				// 	console.log('item_filter_value', item_filter_value);

					let alloptions = Array.from(ddlItems.options);
					console.log('alloptions', alloptions);
					const optionToSelect = alloptions.find(item => item.text ===item_filter_value);
					console.log('optionToSelect', optionToSelect);
					optionToSelect.selected = true;

					//ddlItems.value = item_filter_value;
				}
				else{
					console.log('else item_filter_value');
				}
			}
		}
	});
};

*/

File: rom_app/restaurant_ops_mgmt/report/chef_production_register/chef_production_register.py
import frappe
import yaml
import json


def execute(filters=None):
    print("=========================")
    print(yaml.dump(filters))
    data, columns = [], []
    columns = get_columns()
    cs_data = get_data(filters)

    data = []
    for d in cs_data:
        row = frappe._dict({
            'name': d.name,
            'date': d.date,
            'user_name': d.user_name,
            'branch': d.branch,
            'category': d.category,
            'item': d.item,
            'production_qty': d.production_qty,
            'balance_portion': d.balance_portion,
            'wastage_qty': d.wastage_qty,
            'rate': d.rate,
            'wastage_amount': d.wastage_amount,
        })
        data.append(row)

    return columns, data


def get_columns():
    return [
        {
            'fieldname': 'name',
            'label': 'Id',
            'fieldtype': 'Link',
            'options': 'Chef Production',
        },
        {
            'fieldname': 'date',
            'label': 'Date',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'user_name',
            'label': 'User Name',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'branch',
            'label': 'Branch Name',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'category',
            'label': 'Category',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'item',
            'label': 'Item',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'production_qty',
            'label': 'Prod. Qty',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'balance_portion',
            'label': 'Bal. Qty',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'wastage_qty',
            'label': 'Waste. Qty',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'rate',
            'label': 'Rate',
            'fieldtype': 'Data',
        },
        {
            'fieldname': 'wastage_amount',
            'label': 'Waste. Amt',
            'fieldtype': 'Data',
        }
    ]


@frappe.whitelist()
def get_data(filters):

    conditions = get_conditions(filters)
    print("-------- get data ------------")
    print(conditions)
    build_sql = """
    SELECT
    *
    FROM
    (
        (
    SELECT
        parent1.`name`,
        parent1.`date`,
        parent1.`user_name`,
        parent1.`branch`,
        'Briyani' as category,
        child1.`briyani_category` as item,
        child1.`product_qtykg` as production_qty,
        child1.`balance_portion` as balance_portion,
        child1.`waste_portion` as wastage_qty,
        child1.`rateportion` as rate,
        child1.`wastage_amount` as wastage_amount
    FROM
        `tabChef Production` parent1
    JOIN `tabChef Prod Child Briyani` child1
        ON
        parent1.`name` = child1.`parent`
        )
    UNION
        (
    SELECT
        parent2.`name`,
        parent2.`date`,
        parent2.`user_name`,
        parent2.`branch`,
        'Chicken' as category,
        child2.`chicken_category` as item,
        child2.`product_qtykg` as production_qty,
        child2.`balance_portion` as balance_portion,
        child2.`waste_portion` as wastage_qty,        
        child2.`rateportion` as rate,
        child2.`wastage_amount` as wastage_amount
    FROM
        `tabChef Production` parent2
    JOIN `tabChef Prod Child Chicken` child2
        ON
        parent2.`name` = child2.`parent`
        )
        )
    AS prod
        """
    where_cond = f" WHERE date between '{conditions['from_date_filter']}' AND  '{conditions['to_date_filter']}' "
    if "branch_filter" in conditions:
        where_cond = where_cond + f" AND branch = '{conditions['branch_filter']}' "
    if "category_filter" in conditions:
        where_cond = where_cond + f" AND category = '{conditions['category_filter']}' "
    if "item_filter" in conditions:
        where_cond = where_cond + f" AND item = '{conditions['item_filter']}' "

    order_by = "ORDER BY prod.name ASC, prod.category ASC"

    build_sql = f"{build_sql}  {where_cond} {order_by}"
    print("-------- full sql ------------")
    print(build_sql)
    data = frappe.db.sql(build_sql, as_dict=True)
    return data


def get_conditions(filters):
    conditions = {}
    if (type(filters) is str):
        filters = json.loads(filters)

    for key, value in filters.items():
        if filters.get(key):
            conditions[key] = value
    return conditions


@frappe.whitelist(allow_guest=True)
def get_data_groupby_briyani(filters):
    conditions = get_conditions(filters)
    print("-------- get data ------------")
    print(conditions)
    build_sql = """
    SELECT
    child1.`briyani_category` as item,
    sum(child1.`wastage_amount`) as wastage_amount
    FROM
    `tabChef Production` parent1
    JOIN `tabChef Prod Child Briyani` child1
    ON
    parent1.`name` = child1.`parent`
    """
    where_cond = f" WHERE parent1.date between '{conditions['from_date_filter']}' AND  '{conditions['to_date_filter']}' "
    if "branch_filter" in conditions:
        where_cond = where_cond + f" AND parent1.branch = '{conditions['branch_filter']}' "

    group_by = " GROUP By child1.briyani_category "

    order_by = " ORDER BY item ASC "

    build_sql = f"{build_sql}  {where_cond} {group_by} {order_by}"
    print("-------- full sql ------------")
    print(build_sql)
    data = frappe.db.sql(build_sql, as_dict=True)
    print(data)
    return data


@frappe.whitelist()
def get_data_groupby_chicken(filters):
    conditions = get_conditions(filters)
    print("-------- get data ------------")
    print(conditions)
    build_sql = """
    SELECT
    child2.chicken_category as item, sum(child2.`wastage_amount`) as wastage_amount
    FROM
    `tabChef Production` parent2
    JOIN `tabChef Prod Child Chicken` child2
    ON
    parent2.`name` = child2.`parent`
    """

    where_cond = f" WHERE parent2.date between '{conditions['from_date_filter']}' AND  '{conditions['to_date_filter']}' "
    if "branch_filter" in conditions:
        where_cond = where_cond + f" AND parent2.branch = '{conditions['branch_filter']}' "

    group_by = " GROUP By child2.chicken_category "

    order_by = " ORDER BY item ASC "

    build_sql = f"{build_sql}  {where_cond} {group_by} {order_by}"
    print("-------- full sql ------------")
    print(build_sql)
    data = frappe.db.sql(build_sql, as_dict=True)
    print(data)
    return data
